rflx/pyrflx/typevalue.py

```diff
@@ -727,47 +767,139 @@ def set_refinement(fld: MessageValue.Field, fld_name: str) -> None:
             raise KeyError(f"cannot access field {field_name}")
 
         self.__update_simplified_mapping()
+        check_outgoing_condition_satisfied()
 
-        if all(
-            [
-                self.__simplified(o.condition) == FALSE
-                for o in self._type.outgoing(Field(field_name))
-            ]
-        ):
-            self._fields[field_name].typeval.clear()
-            if isinstance(value, bytes):
-                value_repr = "x" + value.hex()
-            else:
-                value_repr = str(value)
-
-            raise ValueError(
-                f"none of the field conditions "
-                f"{[str(o.condition) for o in self._type.outgoing(Field(field_name))]}"
-                f" for field {field_name} have been met by the assigned value: {value_repr}"
-            )
-
-        self._preset_fields(field_name)
-        self._update_accessible_fields()
+        if checksum_calculation:
+            self._preset_fields(field_name)
+            self._update_accessible_fields()
+            for checksum in self._checksums.values():
+                if self._is_checksum_settable(checksum):
+                    self._calculate_checksum(checksum)
 
     def _preset_fields(self, fld: str) -> None:
         nxt = self._next_field(fld)
         while nxt and nxt != FINAL.name:
             field = self._fields[nxt]
             if not self._has_first(nxt) or not self._has_length(nxt):
                 break
-
             field.first = self._get_first(nxt)
             if isinstance(field.typeval, OpaqueValue):
                 field.typeval.set_expected_size(self._get_length(nxt))
-
             if field.set and isinstance(field.typeval, OpaqueValue):
                 field.first = UNDEFINED
                 field.typeval.clear()
                 break
-
             self._last_field = nxt
             nxt = self._next_field(nxt)
 
+    def set_checksum_function(self, checksum_field_name: str, checksum_method: Callable) -> None:
+        if checksum_field_name not in self.fields:
+            raise KeyError(
+                f"cannot set checksum function: field {checksum_field_name} is not defined"
+            )
+        for field_name, checksum in self._checksums.items():
+            if field_name == checksum_field_name:
+                checksum.set_checksum_function(checksum_method)
+                return
+        raise KeyError(
+            f"cannot set checksum function: field {checksum_field_name} "
+            f"has not been defined as a checksum field"
+        )
```

Comment:
Would it make sense to change the function to `def set_checksum_function(self, checksum: Dict[str, Callable]) -> None` so one doesn't have to call it multiple times if multiple checksums are required?

Reply:
Changed in 74b6ae9